code/modules/vehicles/mecha/equipment/weapons/weapons.dm

```diff
@@ -4,54 +4,216 @@
 	. = ..()
-	firing_effect = new firing_effect
+	equip_cooldown = projectile_delay
+	muzzle_flash = new(src, muzzle_iconstate)
+
+/obj/item/mecha_parts/mecha_equipment/weapon/attach(obj/vehicle/sealed/mecha/M, attach_right)
+	. = ..()
+	AddComponent(/datum/component/automatedfire/autofire, projectile_delay, projectile_delay, projectile_burst_delay, burst_amount, fire_mode, CALLBACK(src, .proc/set_bursting), CALLBACK(src, .proc/reset_fire), CALLBACK(src, .proc/fire))
 
 /obj/item/mecha_parts/mecha_equipment/weapon/action(mob/source, atom/target, list/modifiers)
 	if(!action_checks(target))
 		return FALSE
 	. = ..()
-	chassis.vis_contents += firing_effect
-	fire_bullet(source, target)
-	addtimer(CALLBACK(src, .proc/remove_flash), 2)
-	for(var/i in 1 to projectiles_per_shot)
-		addtimer(CALLBACK(src, .proc/fire_bullet, source, target), (i-1)*projectile_delay)
 
-	chassis.log_message("Fired from [name], targeting [target].", LOG_ATTACK)
+	set_target(get_turf_on_clickcatcher(target, source, list2params(modifiers)))
+	if(!current_target)
+		return
+	if(windup_delay && windup_checked == WEAPON_WINDUP_NOT_CHECKED)
+		windup_checked = WEAPON_WINDUP_CHECKING
+		playsound(chassis.loc, windup_sound, 30, TRUE)
+		if(!do_after(source, windup_delay, TRUE, chassis, BUSY_ICON_DANGER, BUSY_ICON_DANGER, extra_checks = CALLBACK(src, .proc/do_after_checks, current_target)))
+			windup_checked = WEAPON_WINDUP_NOT_CHECKED
+			return
+		windup_checked = WEAPON_WINDUP_CHECKED
+	if(QDELETED(current_target))
+		windup_checked = WEAPON_WINDUP_NOT_CHECKED
+		return
+	current_firer = source
+	if(fire_mode == GUN_FIREMODE_SEMIAUTO)
+		if(!INVOKE_ASYNC(src, .proc/fire) || windup_checked == WEAPON_WINDUP_CHECKING)
+			return
+		reset_fire()
+		return
+	RegisterSignal(source, COMSIG_MOB_MOUSEUP, .proc/stop_fire)
+	RegisterSignal(source, COMSIG_MOB_MOUSEDRAG, .proc/change_target)
+	SEND_SIGNAL(src, COMSIG_MECH_FIRE)
+	source?.client?.mouse_pointer_icon = 'icons/effects/supplypod_target.dmi'
 
-///removes the flash object from viscontents
-/obj/item/mecha_parts/mecha_equipment/weapon/proc/remove_flash()
-	chassis.vis_contents -= firing_effect
+/obj/item/mecha_parts/mecha_equipment/weapon/proc/set_bursting(bursting)
+	if(bursting)
+		ADD_TRAIT(src, TRAIT_GUN_BURST_FIRING, VEHICLE_TRAIT)
+		return
+	REMOVE_TRAIT(src, TRAIT_GUN_BURST_FIRING, VEHICLE_TRAIT)
 
-///proc that actually fires the bullet after timer, should use autofire instead
-/obj/item/mecha_parts/mecha_equipment/weapon/proc/fire_bullet(mob/source, atom/target)
-	if(energy_drain && !chassis.has_charge(energy_drain))//in case we run out of energy mid-burst, such as emp
+///Changes the current target.
+/obj/item/mecha_parts/mecha_equipment/weapon/proc/change_target(datum/source, atom/src_object, atom/over_object, turf/src_location, turf/over_location, src_control, over_control, params)
+	SIGNAL_HANDLER
+	set_target(get_turf_on_clickcatcher(over_object, source, params))
```

Comment:
Maybe rotate mech if needed ?

Reply:
I wanna let players decide when they move for now if its annoying during tm ill add it